File: src/1022_SumofRoot2Leaf.java
class SumofRoot2Leaf {
  int res;
  public int sumRootToLeaf(TreeNode root) {
    res = 0;
    sumRootToLeaf(root, 0);
    return res;
  }

  void sumRootToLeaf(TreeNode root, int val) {
    if (root==null) return;

    val = (val << 1) + root.val;
    if (root.left==null && root.right==null) {
      res += val;
      return;
    }

    sumRootToLeaf(root.left, val);
    sumRootToLeaf(root.right, val);
  }
}

File: src/1022_SumofRoot2Leaf.cpp
#include <local_leetcode.hpp>

class Solution {
  private:
    int dfs(TreeNode*root, int acc) {
      if (root==nullptr) 
        return 0;

      acc = (acc << 1) + root->val;

      // leaf found
      if (root->left==nullptr && root->right==nullptr) 
        return acc;

      return dfs(root->left, acc) + dfs(root->right, acc);
    }

  public:
    int sumRootToLeaf(TreeNode* root) {
      return dfs(root, 0);
    }
};


int main() {
  EXECS(Solution::sumRootToLeaf);
  return 0;
}